File: packages/continuity/src/Restore.ts
import { admissibleForHandoff, type RpcSnapshotEnvelope } from './Envelope.js'
import { allObligations, type RpcObligation, type RpcTimerPolicy } from './Obligations.js'

/**
 * What the successor says it can do with the work the old activation was holding.
 *
 * **The manifest is an observed fact and this is a reviewed policy**, and the design keeps them
 * apart on purpose. The runtime knows that timer `mix-dwell` is active and due in forty seconds.
 * Only the revision being handed to knows what continuing, restarting or firing `mix-dwell` *means* -
 * whether that forty seconds is a dwell somebody is waiting out or a watchdog that should never have
 * been paused. A design that let the first stand in for the second would be inventing plant
 * behaviour at handoff time, which is the moment it is least able to be checked.
 */

/** How one obligation ends up. The design's five, and the difference between them matters. */
export type RpcResolution =
    /** Continuity is preserved under an explicitly supported rule. */
    | 'assumed'
    /** The resource is recreated, with declared delivery and ordering consequences. Recorded as such. */
    | 'reestablished'
    /** Quiescence finished it before the capture, so there is nothing to carry. */
    | 'completed'
    /** The existing Source RPC failure semantics are delivered to whoever is owed them. */
    | 'failed'
    /** The target cannot preserve or safely resolve it, so the handoff is refused. */
    | 'unhonourable'

/** What a revision declares it knows how to do with one obligation id. */
export interface RpcRestoreDeclaration {
    readonly id: string
    readonly resolution: Exclude<RpcResolution, 'unhonourable'>
    /**
     * Required for a timer, and there is no default - see `RpcTimerPolicy`. A timer whose
     * disposition nobody declared refuses the handoff rather than being given a sensible-looking one.
     */
    readonly timerPolicy?: RpcTimerPolicy
    /**
     * For a re-established subscription or publication: whether the transport underneath may
     * duplicate or drop on the way back up, and what is done about it.
     *
     * Required for `reestablished`, because "recreated" without it is a claim of continuity that the
     * transport has not made. A consumer that follows the logical address must not see a false state
     * reset merely because the process changed, and saying how that is achieved is the declaration.
     */
    readonly redelivery?: 'exactly-once' | 'at-least-once-deduplicated' | 'may-drop'
}

export interface RpcRestorePlanEntry {
    readonly id: string
    readonly kind: RpcObligation['kind']
    readonly resolution: RpcResolution
    /** Why, in the words somebody reviewing a handoff has to read. */
    readonly why: string
    readonly timerPolicy?: RpcTimerPolicy
    readonly redelivery?: RpcRestoreDeclaration['redelivery']
}

export type RpcRestorePlan =
    | { readonly admissible: true; readonly entries: readonly RpcRestorePlanEntry[] }
    | { readonly admissible: false; readonly why: string; readonly entries: readonly RpcRestorePlanEntry[] }

/** The clock a timer is measured against, supplied so a plan can be made deterministically. */
export interface RpcRestoreClock {
    readonly now: bigint
}

const timerWhy = (policy: RpcTimerPolicy, obligation: RpcObligation, clock: RpcRestoreClock): { resolution: RpcResolution; why: string } => {
    if (obligation.kind !== 'timer' && obligation.kind !== 'watchdog') return { resolution: 'unhonourable', why: `${obligation.id} is a ${obligation.kind}, and a timer policy says nothing about it` }
    const overdue = obligation.dueAt <= clock.now
    switch (policy) {
        case 'refuse-if-overdue':
            return overdue
                ? { resolution: 'unhonourable', why: `${obligation.id} came due at ${obligation.dueAt} during the handoff, and its declared policy is to abort rather than decide what a missed deadline meant` }
                : { resolution: 'assumed', why: `${obligation.id} is not yet due, so it carries across unchanged` }
        case 'preserve-deadline':
            return { resolution: overdue ? 'failed' : 'assumed', why: overdue ? `${obligation.id} fell due during the handoff and its deadline was preserved, so it is delivered late rather than silently dropped` : `${obligation.id} keeps the deadline it had` }
        case 'preserve-remaining':
            return { resolution: 'assumed', why: `${obligation.id} was effectively paused for the handoff and resumes with the time it had left` }
        case 'restart':
            return { resolution: 'reestablished', why: `${obligation.id} begins its declared duration again, so anything waiting on the original interval waits longer` }
        case 'fire-on-activation':
            return { resolution: 'reestablished', why: `${obligation.id} is delivered as soon as the successor is authoritative, whatever it had left` }
    }
}

/**
 * Whether this snapshot can be restored into this revision, and what happens to each thing it owed.
 *
 * Refuses on the first thing nobody can honour rather than reporting a plan with a hole in it: a
 * partial plan reads as progress, and the one thing a handoff must not do is proceed having
 * quietly not resolved something.
 *
 * **An obligation the target has not declared is `unhonourable`, not `assumed`.** Silence is not
 * consent here - a revision that has never heard of `mix-dwell` cannot be said to have preserved it,
 * and a handoff that treated an unmentioned timer as carried across would hand a plant to a program
 * that does not know it is holding a deadline.
 */
export const planRestore = (snapshot: RpcSnapshotEnvelope, declarations: readonly RpcRestoreDeclaration[], clock: RpcRestoreClock): RpcRestorePlan => {
    const inadmissible = admissibleForHandoff(snapshot)
    if (inadmissible) return { admissible: false, why: inadmissible, entries: [] }

    const declared = new Map(declarations.map((declaration) => [declaration.id, declaration]))
    const entries: RpcRestorePlanEntry[] = []
    let refusal: string | undefined

    for (const obligation of allObligations(snapshot.obligations!)) {
        const declaration = declared.get(obligation.id)
        if (!declaration) {
            const why = `${obligation.id} is a ${obligation.kind} the old activation was holding and this revision declares nothing about it; silence is not a claim to have preserved it`
            entries.push({ id: obligation.id, kind: obligation.kind, resolution: 'unhonourable', why })
            refusal ??= why
            continue
        }

        if (obligation.kind === 'timer' || obligation.kind === 'watchdog') {
            const policy = declaration.timerPolicy
            if (!policy) {
                const why = `${obligation.id} is a ${obligation.kind} and no restore policy was declared for it; there is no default, because every one of them is right for something and catastrophic for something else`
                entries.push({ id: obligation.id, kind: obligation.kind, resolution: 'unhonourable', why })
                refusal ??= why
                continue
            }
            const outcome = timerWhy(policy, obligation, clock)
            entries.push({ id: obligation.id, kind: obligation.kind, resolution: outcome.resolution, why: outcome.why, timerPolicy: policy })
            if (outcome.resolution === 'unhonourable') refusal ??= outcome.why
            continue
        }

        if (obligation.kind === 'lease' && declaration.resolution === 'assumed' && !obligation.issuerSupportsLogicalOwner) {
            const why = `${obligation.id} is a lease from ${obligation.issuer}, which does not support a logical owner; assuming it would hand the successor an authority the issuer does not believe it has`
            entries.push({ id: obligation.id, kind: obligation.kind, resolution: 'unhonourable', why })
            refusal ??= why
            continue
        }

        if (declaration.resolution === 'reestablished' && !declaration.redelivery) {
            const why = `${obligation.id} is declared re-established with no statement about redelivery; "recreated" without one is a claim of continuity the transport underneath has not made`
            entries.push({ id: obligation.id, kind: obligation.kind, resolution: 'unhonourable', why })
            refusal ??= why
            continue
        }

        entries.push({
            id: obligation.id,
            kind: obligation.kind,
            resolution: declaration.resolution,
            why: `${obligation.id} is declared ${declaration.resolution} by this revision`,
            ...(declaration.redelivery ? { redelivery: declaration.redelivery } : {})
        })
    }

    return refusal ? { admissible: false, why: refusal, entries } : { admissible: true, entries }
}

/**
 * The final check, repeated against the snapshot actually taken at the barrier.
 *
 * The design says validation happens four times and this is the last of them, and repeating it here
 * is not belt and braces. The earlier passes ran against a *representative* snapshot - a clone, a
 * dry run, whatever was current when the preparation started - and the thing about to be restored is
 * a different snapshot taken later, with whatever the component did in between. A plan proved
 * against the first says nothing about the second.
 */
export const validateAtBarrier = (
    snapshot: RpcSnapshotEnvelope,
    declarations: readonly RpcRestoreDeclaration[],
    clock: RpcRestoreClock,
    earlier: RpcRestorePlan
): { readonly agreed: true; readonly plan: RpcRestorePlan } | { readonly agreed: false; readonly why: string; readonly plan: RpcRestorePlan } => {
    const plan = planRestore(snapshot, declarations, clock)
    if (!plan.admissible) return { agreed: false, why: plan.why, plan }
    if (!earlier.admissible) return { agreed: false, why: `the earlier pass refused: ${earlier.why}`, plan }

    const before = new Map(earlier.entries.map((entry) => [entry.id, entry.resolution]))
    for (const entry of plan.entries) {
        const was = before.get(entry.id)
        if (was === undefined) return { agreed: false, why: `${entry.id} was not in the snapshot the plan was proved against; the component took on work while the handoff was being prepared`, plan }
        if (was !== entry.resolution) return { agreed: false, why: `${entry.id} was ${was} when the plan was proved and is ${entry.resolution} at the barrier`, plan }
    }
    for (const [id] of before) if (!plan.entries.some((entry) => entry.id === id)) return { agreed: false, why: `${id} was in the plan and is not in the barrier snapshot; it finished while the handoff was being prepared, which changes what the successor is owed`, plan }
    return { agreed: true, plan }
}

File: packages/csharp/SourceRpc.Continuity/Restore.cs
namespace SourceRpc.Continuity;

/// <summary>
/// What the successor says it can do with what the old activation was holding.
///
/// A port of `Restore.ts`, and the rule it exists to keep is the same in both languages: the
/// manifest is an *observed fact* and a declaration is a *reviewed policy*, and one may never stand
/// in for the other. A runtime that let the first substitute for the second would be inventing plant
/// behaviour at the moment it is least able to be checked.
///
/// It matters more here than it did in one language. A .NET revision taking over from a TypeScript
/// one has no compiler in common with it: what it knows about the incumbent's obligations is what
/// the snapshot says, and every disposition it claims is a claim rather than something a type
/// checker agreed to.
/// </summary>
public enum RpcResolution
{
    /// <summary>The successor holds the same obligation, unchanged.</summary>
    Assumed,

    /// <summary>It holds an equivalent one, and something observable differs.</summary>
    Reestablished,

    /// <summary>It was discharged during the handoff.</summary>
    Completed,

    /// <summary>It could not be, and whoever is owed the result is told.</summary>
    Failed,

    /// <summary>Nobody can, and the handoff does not happen.</summary>
    Unhonourable
}

/// <summary>One thing the successor says about one obligation.</summary>
public sealed record RpcRestoreDeclaration
{
    public required string Id { get; init; }
    public required RpcResolution Resolution { get; init; }

    /// <summary>Required for a timer or a watchdog. There is no default, deliberately.</summary>
    public RpcTimerPolicy? TimerPolicy { get; init; }

    /// <summary>Required to re-establish a subscription: what the transport will actually do to it.</summary>
    public RpcRedelivery? Redelivery { get; init; }
}

public sealed record RpcRestorePlanEntry
{
    public required string Id { get; init; }
    public required string Kind { get; init; }
    public required RpcResolution Resolution { get; init; }

    /// <summary>Why, in the words somebody reviewing a handoff has to read.</summary>
    public required string Why { get; init; }

    public RpcTimerPolicy? TimerPolicy { get; init; }
    public RpcRedelivery? Redelivery { get; init; }
}

public sealed record RpcRestorePlan
{
    public required bool Admissible { get; init; }
    public string? Why { get; init; }
    public IReadOnlyList<RpcRestorePlanEntry> Entries { get; init; } = [];
}

/// <summary>The clock a timer is measured against, supplied so a plan can be made deterministically.</summary>
public sealed record RpcRestoreClock
{
    public required long Now { get; init; }
}

public static class RpcRestore
{
    /// <summary>
    /// Pair every obligation with what the successor declares about it, and refuse on the first one
    /// nobody can honour.
    ///
    /// Refusing on the first rather than reporting every failure is deliberate: a partial plan reads
    /// as progress, and the one thing a handoff must not do is proceed having quietly not resolved
    /// something.
    /// </summary>
    public static RpcRestorePlan Plan(RpcSnapshotEnvelope snapshot, IReadOnlyList<RpcRestoreDeclaration> declarations, RpcRestoreClock clock)
    {
        var obligations = snapshot.Obligations;
        if (obligations is null)
            return new RpcRestorePlan { Admissible = false, Why = $"{snapshot.SnapshotId} carries no obligations manifest, so nothing is known about the work the old activation still owed" };

        var declared = declarations.ToDictionary(one => one.Id);
        var entries = new List<RpcRestorePlanEntry>();
        foreach (var (id, kind, decide) in Each(obligations))
        {
            if (!declared.TryGetValue(id, out var declaration))
            {
                // Silence is not consent. A revision that has never heard of this cannot be said to
                // have preserved it, and a handoff that treated an unmentioned timer as carried
                // across would hand a plant to a program that does not know it holds a deadline.
                var entry = new RpcRestorePlanEntry
                {
                    Id = id,
                    Kind = kind,
                    Resolution = RpcResolution.Unhonourable,
                    Why = $"the successor declares nothing about it; silence is not a claim"
                };
                entries.Add(entry);
                return new RpcRestorePlan { Admissible = false, Why = $"{id} ({kind}): {entry.Why}", Entries = entries };
            }
            var resolved = decide(declaration, clock);
            entries.Add(resolved);
            if (resolved.Resolution == RpcResolution.Unhonourable) return new RpcRestorePlan { Admissible = false, Why = $"{id} ({kind}): {resolved.Why}", Entries = entries };
        }
        return new RpcRestorePlan { Admissible = true, Entries = entries };
    }

    /// <summary>
    /// Prove a plan again against the snapshot actually taken at the barrier.
    ///
    /// The earlier pass ran against whatever was current when preparation started. A component that
    /// took on work in between, or finished something, is owed a different set of things - and the
    /// moment before a cutover is the worst possible time to discover it.
    /// </summary>
    public static (bool Agreed, string? Why) ValidateAtBarrier(RpcSnapshotEnvelope atBarrier, IReadOnlyList<RpcRestoreDeclaration> declarations, RpcRestoreClock clock, RpcRestorePlan earlier)
    {
        var now = Plan(atBarrier, declarations, clock);
        if (!now.Admissible) return (false, now.Why);
        var before = earlier.Entries.Select(entry => entry.Id).ToHashSet(StringComparer.Ordinal);
        var after = now.Entries.Select(entry => entry.Id).ToHashSet(StringComparer.Ordinal);
        var arrived = after.Except(before).ToList();
        if (arrived.Count > 0) return (false, $"{string.Join(", ", arrived)} took on work while the handoff was being prepared, and the plan proved earlier says nothing about it");
        var gone = before.Except(after).ToList();
        if (gone.Count > 0) return (false, $"{string.Join(", ", gone)} finished while the handoff was being prepared, so the successor is owed a different set of things than the plan proved earlier");
        return (true, null);
    }

    private static IEnumerable<(string Id, string Kind, Func<RpcRestoreDeclaration, RpcRestoreClock, RpcRestorePlanEntry> Decide)> Each(RpcObligations obligations)
    {
        foreach (var one in obligations.Timers) yield return (one.Id, "timer", (declaration, clock) => Timer(one.Id, "timer", one.DueAt, declaration, clock));
        foreach (var one in obligations.Watchdogs) yield return (one.Id, "watchdog", (declaration, clock) => Timer(one.Id, "watchdog", one.DueAt, declaration, clock));
        foreach (var one in obligations.OutboundCalls) yield return (one.Id, "outbound-call", (declaration, _) => Simple(one.Id, "outbound-call", declaration));
        foreach (var one in obligations.InboundWork) yield return (one.Id, "inbound-work", (declaration, _) => Simple(one.Id, "inbound-work", declaration));
        foreach (var one in obligations.Subscriptions) yield return (one.Id, "subscription", (declaration, _) => Subscription(one.Id, declaration));
        foreach (var one in obligations.PendingPublications) yield return (one.Id, "publication", (declaration, _) => Simple(one.Id, "publication", declaration));
        foreach (var one in obligations.Leases) yield return (one.Id, "lease", (declaration, _) => Lease(one, declaration));
        foreach (var one in obligations.Sequences) yield return (one.Id, "sequence", (declaration, _) => Simple(one.Id, "sequence", declaration));
    }

    private static RpcRestorePlanEntry Timer(string id, string kind, long dueAt, RpcRestoreDeclaration declaration, RpcRestoreClock clock)
    {
        if (declaration.TimerPolicy is null)
            return new RpcRestorePlanEntry
            {
                Id = id,
                Kind = kind,
                Resolution = RpcResolution.Unhonourable,
                // No default, because every policy is right for something and catastrophic for
                // something else, and nobody can pick without knowing what this timer is for.
                Why = "no restore policy was declared for it, and there is no default: a dwell that restarts has doubled a bake, and a watchdog that preserved its deadline fires the instant the successor comes up"
            };
        var overdue = dueAt <= clock.Now;
        var policy = declaration.TimerPolicy.Value;
        return policy switch
        {
            RpcTimerPolicy.PreserveRemaining => Entry(id, kind, RpcResolution.Assumed, "the handoff is not counted against it: it fires after the time it had left", policy),
            RpcTimerPolicy.PreserveDeadline when overdue =>
                Entry(id, kind, RpcResolution.Failed, "its deadline passed during the handoff and it preserves deadlines, so it is delivered late rather than silently dropped - a deadline that was real is a fact somebody is owed", policy),
            RpcTimerPolicy.PreserveDeadline => Entry(id, kind, RpcResolution.Assumed, "it fires when it would have fired", policy),
            RpcTimerPolicy.Restart => Entry(id, kind, RpcResolution.Reestablished, "it begins its declared duration again, which is observably not the timer that was running", policy),
            RpcTimerPolicy.FireOnActivation => Entry(id, kind, RpcResolution.Reestablished, "it is delivered as soon as the successor is authoritative, which is not when it was due", policy),
            RpcTimerPolicy.RefuseIfOverdue when overdue =>
                Entry(id, kind, RpcResolution.Unhonourable, "its deadline passed while the handoff was being prepared, and it is declared to abort rather than decide what a missed deadline meant", policy),
            RpcTimerPolicy.RefuseIfOverdue => Entry(id, kind, RpcResolution.Assumed, "it is not yet due, so there is nothing to refuse", policy),
            _ => Entry(id, kind, RpcResolution.Unhonourable, $"the policy {policy} is not one this implementation knows", policy)
        };
    }

    private static RpcRestorePlanEntry Subscription(string id, RpcRestoreDeclaration declaration)
    {
        if (declaration.Resolution == RpcResolution.Reestablished && declaration.Redelivery is null)
            return new RpcRestorePlanEntry
            {
                Id = id,
                Kind = "subscription",
                Resolution = RpcResolution.Unhonourable,
                Why =
                    "it is declared re-established without saying what the transport will do to it, which is a claim of continuity the transport underneath has not made - and a consumer following the logical address must not see a false reset because a process changed"
            };
        return new RpcRestorePlanEntry
        {
            Id = id,
            Kind = "subscription",
            Resolution = declaration.Resolution,
            Why = declaration.Redelivery is { } redelivery ? $"re-established with {RpcNames.Redelivery(redelivery)} redelivery" : "carried across unchanged",
            Redelivery = declaration.Redelivery
        };
    }

    private static RpcRestorePlanEntry Lease(RpcLeaseObligation lease, RpcRestoreDeclaration declaration)
    {
        if (declaration.Resolution == RpcResolution.Assumed && !lease.IssuerSupportsLogicalOwner)
            return new RpcRestorePlanEntry
            {
                Id = lease.Id,
                Kind = "lease",
                Resolution = RpcResolution.Unhonourable,
                Why =
                    $"{lease.Issuer} does not support a logical owner, so a lease it granted cannot be carried across a process change - assuming it would hand the successor an authority the issuer does not believe it has"
            };
        return new RpcRestorePlanEntry { Id = lease.Id, Kind = "lease", Resolution = declaration.Resolution, Why = $"declared {declaration.Resolution.ToString().ToLowerInvariant()} against {lease.Issuer}" };
    }

    private static RpcRestorePlanEntry Simple(string id, string kind, RpcRestoreDeclaration declaration) =>
        new()
        {
            Id = id,
            Kind = kind,
            Resolution = declaration.Resolution,
            Why = $"declared {declaration.Resolution.ToString().ToLowerInvariant()} by the successor"
        };

    private static RpcRestorePlanEntry Entry(string id, string kind, RpcResolution resolution, string why, RpcTimerPolicy policy) =>
        new()
        {
            Id = id,
            Kind = kind,
            Resolution = resolution,
            Why = why,
            TimerPolicy = policy
        };
}
